Fixes: #3862; Added an example Python Support case study

himanshumahajan138 opened the pull request:
# Pull Request 
Added an example Python Support case study
Fixes: #3862 

## Description
Added `Python` Support for Mill From basic installation to multiple supports like virtual environment, mypy integration , pex integration

## Related Issues
- Link to related issue #3862.

## Checklist
- [x] Python-hello example
- [x] Python-modules example
- [x] Python-dependencies example
- [x] Python-mypy support example
- [x] Python-pex Support example
- [x] Documentation Updated

## Additional Notes: (Code Copy and Plagiarism is Strictly Prohibited in case of Bounty)
> **Reminder**: **Contributions must be original or appropriately attributed. **Strict action** may be taken by the community for  cheating and coping the code from some PR for Bounty or plagiarized code to maintain integrity within the community.**

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,193 @@
+/**
+ * A trait for managing Python environments and dependencies in a Mill build.
+ */
+trait PythonModule extends Module {
```
This first example should not use a `trait`, and every defined `Task` should be explained. See the Typescript example for reference

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,193 @@
+ */
+trait PythonModule extends Module {
+
+  /** Specifies dependencies to install in the virtual environment. */
+  def dependencies: T[Seq[String]] = Task { Seq.empty[String] }
+
+  /** Specifies command-line options to pass to the main Python file. */
+  def options: T[Seq[String]] = Task { Seq.empty[String] }
+
+  /** Specifies the name of the virtual environment directory. */
+  def venvName: T[String] = Task { "venv" }
+
```
First example can get-rid-of/hardcode all of these for simplicity

himanshumahajan138 replied:
You mean no need for virtual env and all just installation of pip and then run simple python file 👍

lihaoyi replied:
Yes, they can be added in the later examples when we introduce resolving dependencies

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  /** Executes the main Python file using the virtual environment. */
+  def compile: T[Unit] = Task {
```
Why is this called `compile` and why is it a `Task`?

himanshumahajan138 replied:
silly me i have to give some output like PathRef or something else

actually i have edited this command to return PathRef for pex example but forgot this 

is there any other consideration?

himanshumahajan138 replied:
i think sir you are referring to have one run function which will run the file and compile will only compile all the sources and resources 

am i right ?

lihaoyi replied:
I don't know, you have to decide, but what you have written now between the comment and the method is 100% wrong

himanshumahajan138 replied:
Ok sir, first i will re-search for it in already present examples and then will update it as per other and follow mill structure 👍

lihaoyi commented:
Please clean this up to follow the Typescript example documentation. I cannot fix every detail for you, you need to go and real through those examples yourself and fix your code to match

himanshumahajan138 commented:
Ok @lihaoyi  sir, Thanks for reviewing 

My silly mistake made one test fail its just f string error 😔

No worries i will update the code and documentation

himanshumahajan138 commented:
@lihaoyi Sir Final Testing has been done...

- Documentation Updated (Written by me 😅)
- Code Logic Updated as per TypeScript Example
- combined moduleDeps, virtualEnv & libraryDeps (coz they are dependent to each other)
- All Tests Passed and Code is checked thrice...

Please pardon me if anything missed or done in wrong way...

himanshumahajan138 commented:
@lihaoyi Sir Final Testing has been done...

- Documentation Updated (Written by me 😅)
- Code Logic Updated as per TypeScript Example
- combined moduleDeps, virtualEnv & libraryDeps (coz they are dependent to each other)
- All Tests Passed and Code is checked thrice...

Please pardon me if anything missed or done in wrong way...

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,138 @@
+  val pipPath = os.home / ".local" / "bin" / "pip3"
+
+  if (!os.exists(pipPath)) {
+    os.call(Seq("wget", "-q", "https://bootstrap.pypa.io/get-pip.py", "-O", s"$pipFile"))
```
We should not need to install pip because Python3 bundles pip for all versions >3.4 and above. Let's just use the bundled one

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,138 @@
+// The `compile` Task  ensures `pip` is installed, retrieves the main file directory, and returns it for further processing.
+// This task is considered as a part of code which is responsible for all compilation of sources and pre-requirements.
+
+def compile: T[PathRef] = Task {
+  installPip()
+  val mainFileDir = sources().path
+  
+  PathRef(mainFileDir)
+}
```
We should perform MyPy typecheck validation here, as mentioned in the original ticket

lihaoyi replied:
Maybe rename it `typecheck` instead of `compile`

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,138 @@
+
+def sources: T[PathRef] = Task.Source(millSourcePath / "src")
+
+def allSources: T[Seq[PathRef]] = Task {
```
`allSources` is not used and should be removed

lihaoyi commented on example/extending/newlang/7-python-module-deps/build.mill:
```diff
@@ -0,0 +1,263 @@
+
+  /** Path to the requirements file, configurable by the user. */
+  def requirementsPath: T[PathRef] = Task {
+    PathRef(millSourcePath / "src" / "requirements.txt")
```
This should not be in `src`, just `millSourcePath / "requirements.txt"` is fine

himanshumahajan138 replied:
i think you mean we should have only single `requirements.txt` in the project like if user have multiple sources then also
they have to mention all the requirements in the same file and that will be main requirements for the project

lihaoyi commented on example/extending/newlang/7-python-module-deps/build.mill:
```diff
@@ -0,0 +1,263 @@
+    installPip()
+    val venvDir = T.dest / venvName()
+
+    os.call(Seq(
```
`venv` is bundled with Python3 so we should not need to install it

lihaoyi commented on example/extending/newlang/7-python-module-deps/build.mill:
```diff
@@ -0,0 +1,263 @@
+  }
+
+  /** Activates the created virtual environment. */
+  def activateVenv: T[PathRef] = Task {
```
We should not need to activate the venv, we should directly use the `bin/python` binary from within it

lihaoyi commented on example/extending/newlang/8-python-mypy/build.mill:
```diff
@@ -0,0 +1,225 @@
+// == Python MyPy Support
```
Let's follow the typescript setup: MyPy support should be included in the first example. The last example should be for third-party dependencies and bundling

himanshumahajan138 commented:
@lihaoyi Sir Let me tell you one thing that i have seen actually i  installed Ubuntu in WSL and when i run it for the first time it just only provides python which is installed with only setuptools and nothing else (no pip , no venv and others )

So when i searched on internet i came to know that we have to install pip and venv using `sudo apt install python3-pip` and `sudo apt install python3-venv`
 
but now here comes the problem of permissions we cant get sudo for every task and even this is not the good practise 

so to tackle this problem i have  used official guide for installing pip for users without requiring sudo permission and this will be installed for the user rather than the sudo 

i think this is a good approach, rest if you say i will follow your way...

lefou commented:
A Mill build process should not install anything outside of the project directory, unless it's some kind of cache like `coursier`. We should just depend on installed tools and provide meaningful error messages if they are missing. We can provide installation/setup support in some external module, e.g. `build.PythonModule/installPip`, but that should be purely optional.

himanshumahajan138 commented:
> A Mill build process should not install anything outside of the project directory, unless it's some kind of cache like `coursier`. We should just depend on installed tools and provide meaningful error messages if they are missing. We can provide installation/setup support in some external module, e.g. `build.PythonModule/installPip`, but that should be purely optional.

But Sir what about github actions tests ?

himanshumahajan138 commented:
Sir Please see this image i just installed the ubuntu-24.04 version and when i tried these commands you can see that it don't comes with pip and venv 

**Moreover, i am not installing anything outside the T.dest actually i just downloaded `get-pip.py` in T.dest and then using this we installed pip for user originally that's the official way** 

![image](https://github.com/user-attachments/assets/83b1ecf8-4cc5-4458-a278-55d57c5743e9)

lihaoyi commented:
You should be able to use `python3 -m pip` and `python3 -m venv` without needing installation

himanshumahajan138 commented:
ok then let me recall the updates again:

- no need to install pip and virtualenv
- mypy will be added in first example (5-hello-python) change in compile function
- no need for activating venv
- no need for all sources
- and other minor updates...

Thanks For Review Sir (This community is filled with helpful members❤️)

lihaoyi commented:
@himanshumahajan138 yep! I expect this will take multiple rounds of back and forth review to polish into shape, but it's moving in the right direction so I think we'll eventually get there

himanshumahajan138 commented:
> You should be able to use `python3 -m pip` and `python3 -m venv` without needing installation

Hat's off to your Experience Sir!  You said it should run, and it did 


One More question than i will Move to documentation part :

> Let's follow the typescript setup: MyPy support should be included in the first example. The last example should be for third-party dependencies and bundling

As per this you mean we should add mypy in first example i.e 5-hello-python but for this we need to install mypy dependency and this needs virtual env but we have introduced virtual env in example 7-python-module-deps 

so what should i do now ???

lihaoyi commented:
@himanshumahajan138 we need pip internally to install mypy in the first examples, but we do not need to expose it to the user to install their own things until the later examples. This is similar to the Typescript examples which use NPM from the start but only let users install stuff using NPM later

lihaoyi commented:
> oh so you mean we will download mypy internally without venv into main python on the system and then we will use it and we will type check that in all the other examples as a function named as type checking rather than compile

no, you should use venv

himanshumahajan138 commented:
@lihaoyi Sir, Code and Documentation is Updated and Ready for Review...

Please Pardon me if I made any mistake or missed something...

Thanks...

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,158 @@
+// == Basic Python Pipeline 
+
+// This Example demonstrates integration of https://www.python.org[Python]
+// compilation into a Mill build to compile https://www.python.org[Python] Scripts. Mill
```
As a rule of thumb, you should not link the same keyword more than once per page

himanshumahajan138 replied:
i thought a lot for choosing name 😅

finally i choosed `Python Integration in Mill`

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,158 @@
+  val venvDir = T.dest / "venv"
+
+  os.call(
+    Seq(
```
Don't need the `Seq` Here, don't need the `.toString`, and the `call` should be a one-liner

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,158 @@
+  PathRef(venvDir)
+}
+
+def setup: T[PathRef] = Task {
```
Let's combine this with `createVenv` above

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
````diff
@@ -0,0 +1,158 @@
+  PathRef(pythonVenv)
+}
+
+// At this point, we have a minimal working build, with a build graph that looks like this:
+//
+// ```graphviz
+// digraph G {
+//   rankdir=LR
+//   node [shape=box width=0 height=0 style=filled fillcolor=white]
+//   createVenv -> setup -> typeCheck
+//   sources -> typeCheck
+//   mainFileName -> typeCheck
+// }
+// ```
````
We can probably skip this diagram since it's not yet interesting; the one below is enough

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,158 @@
+  val pythonVenv = typeCheck().path
+  
+  os.call(
+    Seq(
+      pythonVenv.toString,
+      s"${sources().path / mainFileName()}"
+    ) ++ args.value,
```
No need `Seq`, `toSdtring`, or the `s"${}"` wrapper

lihaoyi commented:
@himanshumahajan138 took a pass, most of the comments apply to all the examples.

Getting pretty close I think, looks a lot better than when we started!

himanshumahajan138 commented:
@lihaoyi sir, like i have used type checking in evey example is that a good and right way

lihaoyi commented:
Yes having typechecking in every example is right. The examples are meant to build upon each other, so we would expect them to get longer and more complex as it progresses down the list

himanshumahajan138 commented:
@lihaoyi Sir, finally I have Updated the Code as per reviews and now I think we can move further in review or merging part

Hope you Like it...

Thanks for Reviews Sir...

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,118 @@
+
+// === Type Checking
+
+// The `mainFileName` Task provides the name for the main python script file for running and 
+// The `typeCheck` Task  ensures that the python setup is set and Checks the main file for error free code.
+// This task is considered as a part of code which is responsible for Error Checking of the Main Script 
+// and setting up setup for python script runable environment.
+
+def mainFileName: T[String] = Task { "main.py" }
```
We should move `mainFileName` definition and docs into the same section as `run`, since it is only needed later and is not necessary for `typeCheck`

himanshumahajan138 replied:
done

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,118 @@
+// then we have `main.py` and running this file will result in either a generic greeting 
+// or a personalized one based on the command-line arguments passed, allowing for flexible interaction with the user.
+
+/** See Also: src/main.py */
```
This snippet is far too verbose. We need something 5 lines max

himanshumahajan138 replied:
i tried my best and converted to 10 lines 😅

lihaoyi commented on example/extending/newlang/5-hello-python/build.mill:
````diff
@@ -0,0 +1,118 @@
+// Note that we use `stdout = os.Inherit` since we want to display any output to the user,
+// rather than capturing it for use in our command.
+
+// ```graphviz
````
This diagram is great, but we should have another diagram with just `typeCheck` and its upstream targets above under the `Type Checking` section

himanshumahajan138 replied:
is this good which we have before...
![image](https://github.com/user-attachments/assets/7c69f094-4302-48a2-b3ef-890264fb50c0)

himanshumahajan138 replied:
done

lihaoyi commented on example/extending/newlang/6-python-modules/build.mill:
```diff
@@ -0,0 +1,138 @@
+// This Example demonstrates the use of a `PythonModule` trait to manage Python Scripts within
+// different objects in a Mill build.
+
+// ==== Using Same PythonModule for Mutiple Objects
```
`Mutiple` should be `Multiple`

himanshumahajan138 replied:
done

lihaoyi commented on example/extending/newlang/6-python-modules/build.mill:
```diff
@@ -0,0 +1,138 @@
+
+// ==== Using Same PythonModule for Mutiple Objects
+
+// The `PythonModule` trait simplifies Python project management within the `Mill` build tool by providing essential
+// functionalities such as creating `virtual environment`, managing `source files`, `Checking code for Error`, and `executing Python` scripts.
+// It ensures that users have a streamlined workflow by automating the setup of necessary tools 
+// and organizing project files effectively. 
```
This whole paragraph is too repetitive, since we already discuss this above. We should discuss how `trait PythonModule` takes the same tasks we defined above and lets them be re-used for separate modules

himanshumahajan138 replied:
removed

lihaoyi commented on example/extending/newlang/6-python-modules/build.mill:
```diff
@@ -0,0 +1,138 @@
+package build
+import mill._
+
+/** PythonModule Trait for Basic Python Pipeline in Mill Build Tool */
```
~All these comments are useless and should be removed

himanshumahajan138 replied:
only new things have comments now

lihaoyi commented on example/extending/newlang/6-python-modules/build.mill:
```diff
@@ -0,0 +1,138 @@
+
+// ==== Example
+
+// The script defines three key `objects` that extend the `PythonModule` trait, each object encapsulating a distinct runtime configuration.
+// By allowing objects like `foo`, `bar`, and `qux` to extend the `PythonModule`, it promotes code reuse and modular design, 
+// enabling easy access to common functionalities while facilitating customization for specific needs.
```
This paragraph is bad and meaningless, as are the comments inside the below code block. Just say that we can now instantiate `PythonModule` three times to work with code in `foo/src/`, `foo/bar/src/`, and `qux/src/`

himanshumahajan138 replied:
removed

lihaoyi commented on example/extending/newlang/6-python-modules/build.mill:
```diff
@@ -0,0 +1,138 @@
+
+// We have used three different Python Scripts `foo/src/foo.py`, `foo/bar/src/bar.py`, `qux/src/qux.py` 
+
+/** See Also: foo/src/foo.py */
+/** See Also: foo/bar/src/bar.py */
+/** See Also: qux/src/qux.py */
```
Again, these examples are far too long, and anyway we don't need to include them here since we already saw the file in the previous section

himanshumahajan138 replied:
i shortened them

lihaoyi commented:
Took a review pass. Many of the comments apply to the following docs and examples as well

himanshumahajan138 commented on example/extending/newlang/5-hello-python/build.mill:
```diff
@@ -0,0 +1,118 @@
+
+*/
+
+// Uncomment the Result variable line(Line Number:13) in the `src/main.py` code to see `typeCheck` error
```
actually sir i want to show error also but the problem is that if i write code for error then it will fail in ci test as this will me a major error from mypy side so how can i do this 

please Guide me a little Bit Sir...

himanshumahajan138 commented:
@lihaoyi Sir, Ready for Next Reviews 

what i changed in last updates:
- updated documentation and some code 
- updated scripts and shorten them 
- rewrite the documentation
- removed irrelevant and repeative codes and docs

Thanks For Reviews Sir...

lihaoyi commented:
@himanshumahajan138 I think this looks good. Just need to run `./mill mill.scalalib.scalafmt.ScalafmtModule/`  to reformat everything and we can merge it

himanshumahajan138 commented:
Actually sir before ci don't run for .mill build files but now it do for these files also and its was my fault that i didn't checked it...
No worries let me update and then finalize it...

Thanks sir for your guidance and reviews...

himanshumahajan138 commented:
AND HERE WE GO🎉

Everything Ready for Merge Lets make this happen...

Thanks A Lot Sir ❤️
(To be Honest You have Changed my Life I was Nothing before contributing to Mill and now I am Confident with my Skills and alot financially also)
